pkg/nsm/apis/internsm/internsm.proto

```diff
@@ -0,0 +1,105 @@
+    //  id of the remote connection - must be unique when fully specified
+    RemoteConnectionId id = 1;
+    //  state of the remote connection
+    RemoteConnectionState state = 2;
+    //  network_service_name - Name of the Network Service at the destination of the RemoteConnection
+    string network_service_name = 3;
+    //  network_service_endpoint_name - Name of the specific Network Service Endpoint at the destination of the RemoteConnection
+    string network_service_endpoint_name = 4;
+    //  remote_mechanism - RemoteMechanism for the remote connection.  Must be empty in RequestRemoteConnection.  Must by fully specified in RemoteConnectionReply
+    common.RemoteMechanism remote_mechanism = 5;
+    // connection_context - conntection_context of the remote connection.
+    common.ConnectionContext  connection_context = 6;
+    // labels - meta data about the RemoteConnection.  Typically provided in RequestRemoteConnection and not changed from there.
+    map<string,string> labels = 7;
+}
+
+// RemoteConnectionRequest - Request for a RemoteConnection
+message RemoteConnectionRequest {
+   // remote_connection - partially specified RemoteConnection being requested
+   RemoteConnection remote_connection = 1;
+   // supported_remote_mechanisms - preference ordered list of supported_remote_mechanisms for the source NSM
+   // The destination NSM should try to pick the highest prference remote_mechanism to fully specify and use to complete the connection.
+   repeated common.RemoteMechanism supported_remote_mechanisms = 4;
+}
+
+// RemoteConnectionReply - Reply to a RemoteConnectionRequest
+message RemoteConnectionReply {
+    // success - true of operation was successful - false otherwise
+    // In RequestRemoteConnection this means the destination NSMs end of the RemoteConnection is fully operational
+    // In UpdateRemoteConnection this means the destination NSM has been able to negotiate successfully the
+    //    update to the ConnectionContext with the NSE
+    // In DeleteRemoteConnection this means the connection is now fully deleted in the Destination NSM
+    bool success = 1;
+    // extended_error - String explaining error condition of success == false.  Must be empty if success = true.
+    string extended_error = 2;
+    //  remote_connection - fully specified RemoteConnection
+    RemoteConnection remote_connection = 3;
+}
+
+// RemoteConnectionUpdate - Update on a RemoteConnection
+// Should be returned from MonitorRemoteConnection when:
+//     1)  RemoteConnectionState of the RemoteConnection changes
+//     2)  Periodically to update metrics on RemoteConnection
+// Uses:
+//     - Updating 'recoverying' peer NSM on peer NSM restart on existing RemoteConnections
+//     - Updating peer NSM of unrequested CLOSING of a RemoteConnection
+//     - Updating peer NSM of metrics on RemoteConnections - can be used to detect failure modes requiring
+//       auto-healing
+message RemoteConnectionUpdate {
```

Comment:
RemoteConnection could be used for all type of RPCs. The code would just check fields it needs.

Reply:
@sbezverk metrics are additional information *about* the RemoteConnection, not part of it.  Once established, a RemoteConnection shouldn't 'evolve' except if explicitly 'updated'.  If you put metrics in it, it evolves *constantly*.

Reply:
What are the endpoints of this rpc?

Reply:
RemoteConnectionUpdate isn't an RPC, its a message.  It gets streamed back from MonitorConnections.

So one NSM calls MonitorConnections on another, and gets back a stream of RemoteConnectionUpdate messages that communicate the RemoteConnection's that match the MonitorConnections request, and any metrics associated with those connections.